codec-sockjs/src/main/java/io/netty/handler/codec/sockjs/transport/WebSocketHAProxyHandshaker.java

```diff
@@ -0,0 +1,231 @@
+
+        // Create the WebSocket handshake buildResponse.
+        FullHttpResponse res = new DefaultFullHttpResponse(HTTP_1_1, new HttpResponseStatus(101,
+                isHixie76 ? "WebSocket Protocol Handshake" : "Web Socket Protocol Handshake"));
+        if (headers != null) {
+            res.headers().add(headers);
+        }
+
+        res.headers().add(UPGRADE, WEBSOCKET);
+        res.headers().add(CONNECTION, HttpHeaderValues.UPGRADE);
+
+        // Fill in the headers and contents depending on handshake getMethod.
+        if (isHixie76) {
+            // New handshake getMethod with a challenge:
+            res.headers().add(SEC_WEBSOCKET_ORIGIN, req.headers().get(ORIGIN));
+            res.headers().add(SEC_WEBSOCKET_LOCATION, uri());
+            String subprotocols = req.headers().get(SEC_WEBSOCKET_PROTOCOL);
+            if (subprotocols != null) {
+                String selectedSubprotocol = selectSubprotocol(subprotocols);
+                if (selectedSubprotocol == null) {
+                    throw new WebSocketHandshakeException("Requested subprotocol(s) not supported: " + subprotocols);
+                } else {
+                    res.headers().add(SEC_WEBSOCKET_PROTOCOL, selectedSubprotocol);
+                }
+            }
+
+            // Calculate the answer of the challenge.
+            key1 = req.headers().get(SEC_WEBSOCKET_KEY1);
+            key2 = req.headers().get(SEC_WEBSOCKET_KEY2);
+        } else {
+            // Old Hixie 75 handshake getMethod with no challenge:
+            res.headers().add(WEBSOCKET_ORIGIN, req.headers().get(ORIGIN));
+            res.headers().add(WEBSOCKET_LOCATION, uri());
+            String protocol = req.headers().get(WEBSOCKET_PROTOCOL);
+            if (protocol != null) {
+                res.headers().add(WEBSOCKET_PROTOCOL, selectSubprotocol(protocol));
+            }
+        }
+        return res;
+    }
+
+    protected ByteBuf calculateLastKey(final ByteBuf content) {
+        final int a = (int) (Long.parseLong(BEGINNING_DIGIT.matcher(key1).replaceAll("")) / BEGINNING_SPACE
+                .matcher(key1).replaceAll("").length());
+        final int b = (int) (Long.parseLong(BEGINNING_DIGIT.matcher(key2).replaceAll("")) / BEGINNING_SPACE
+                .matcher(key2).replaceAll("").length());
+        final long c = content.readLong();
+        final ByteBuf input = Unpooled.buffer(16);
+        input.writeInt(a);
+        input.writeInt(b);
+        input.writeLong(c);
+        final ByteBuf key = Unpooled.buffer().writeBytes(md5(input.array()));
+        input.release();
+        return key;
+    }
+
+    @Override
+    public ChannelFuture handshake(Channel channel, FullHttpRequest req) {
+        final FullHttpResponse response = newHandshakeResponse(req, null);
+        return channel.writeAndFlush(response).addListener(new ChannelFutureListener() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 26 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1188)